Rewrite ModificationExecutor

markuspf opened the pull request:
This pull-request is a rewrite of `ModificationExecutor` and surrounding classes. It provides new Insert, Remove, Replace, Update, and Upsert executors.

This rewrite was triggered by the introduction of `ShadowRows` as part of the Subquery optimisation project currently undertaken by Team Gordo, and us wanting to remove the `SingleBlockFetcher`.

This is a draft, at the very least the follwing things remain to do

 - [x] Handle Statistics
 - [x] Spread some \TRI_ASSERTS` around the code
 - [x] Look at writing some GoogleTests for the new classes
 - [x] Actually hook up the new Modifiers (at the moment they should be swappable with the old ones)
 - [ ] Benchmarks to look outi whether any slowdowns are caused by the new code.
 - [x] Address all TODOs in the code

mchacki commented on arangod/Aql/InsertModifier.cpp:
```diff
@@ -0,0 +1,74 @@
+Result InsertModifierCompletion::accumulate(InputAqlItemRow& row) {
+  RegisterId const inDocReg = _modifier._infos._input1RegisterId;
+
+  // The document to be UPDATEd
```
```suggestion
  // The document to be INSERTed
```

mchacki commented on arangod/Aql/ModificationExecutor.h:
```diff
@@ -183,43 +183,26 @@ struct ModificationExecutorInfos : public ExecutorInfos {
+  friend struct Remove;
+  friend struct Upsert;
+  friend struct UpdateReplace<Update>;
+  friend struct UpdateReplace<Replace>;
```
lol names.

However is ok like this ;)

mchacki commented on arangod/Aql/ModificationExecutor2.cpp:
```diff
@@ -0,0 +1,286 @@
+                                                                        Infos& infos)
+    : _infos(infos), _fetcher(fetcher), _modifier(infos) {
+  // important for mmfiles
+  // TODO: Why is this important for mmfiles?
```
In MMFiles we need to make sure that the data is not moved in memory or collected for this collection as soon as we start writing to it.
This pin makes sure that all of the memory moving is not executed and all Pointers we get from a collection stay correct until we release this pin

mchacki commented on arangod/Aql/ModificationExecutor2.cpp:
```diff
@@ -0,0 +1,286 @@
+
+  _modifier.reset();
+
+  size_t maxOutputs = std::min(output.numRowsLeft(), ExecutionBlock::DefaultBatchSize());
```
You can rely on numRowsLeft() here.
The outer code makes sure that output is not larger then DefaultBatchSize()

mchacki commented on arangod/Aql/ModificationExecutor2.cpp:
```diff
@@ -0,0 +1,286 @@
+  // TODO:
+  // We want to say "has more" if upstream still has rows.
+  // sooo... just return the upstream state?
+  return {ExecutionState::DONE, std::move(stats)};
```
You have use the `fetchRow()` command above.
This returns a state.
Here you need to return the last state this fetchRow has returned.

mchacki commented on arangod/Aql/ModificationExecutor2.h:
```diff
@@ -0,0 +1,100 @@
+// Replace and Upsert executors.
+//
+// The fetcher and modification-specific code is spliced in via a template for
+// performance reasons (TODO: verify that this is true).
```
```suggestion
// performance reasons
```

mchacki replied:
It is

mchacki commented on arangod/Aql/ModificationExecutor2.h:
```diff
@@ -0,0 +1,100 @@
+// of Insert and Update/Replace, and hence more complicated. (TODO: Is there any
+// way in which we can redesign the execution of AQL queries that we can just
+// transform an upsert into an update and an insert and get rid of the upsert
+// executor?)
```
Regarding TODO:
Not for the time being.
We do not have alternative execution paths

ObiWahn commented on arangod/Aql/SimpleModifier.cpp:
```diff
@@ -0,0 +1,157 @@
+  _accumulator.clear();
+  _operations.clear();
+  _results = OperationResult{};
+  _accumulator.openArray();
```
Please do not open and close in different places!

goedderz commented:
Good idea @ObiWahn.

markuspf commented:
> Please use the RAII idiom when opening and closing builders. It is really hard to reason about when they are opened and closed in arbitrary locations.

Note that this is not quite as easy as one might like: Scoping with a guard will not work, because the accumulator needs to be able to stay alive across multiple calls to `produceRows`, so we can only do this with object lifetime; Hence this PR now contains a tasty hack that wraps a VPackBuilder inside an object so that we can just create a new accumulator when we need one, and can only add slices (and never close the array which is opened on creation, and closed when the contents of the accumulator are observed. Quantum supremacy.)

markuspf commented:
This is now being tested here: http://jenkins.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/6921/ fingers crossed.

ObiWahn commented on arangod/Aql/ModificationExecutorHelpers.cpp:
```diff
@@ -0,0 +1,224 @@
+
+  // If `value` is a string, this is our _key entry, so we use that.
+  if (value.isString()) {
+    key.assign(value.slice().copyString());
```
Could be string_view soon https://github.com/arangodb/arangodb/pull/10287 -- no action needed

ObiWahn replied:
Already perfect as `assign` takes `&&`

ObiWahn commented on tests/Aql/InsertExecutorTest.cpp:
```diff
@@ -312,12 +312,15 @@ TEST_P(InsertExecutorTestCount, insert_with_key_and_overwrite) {
     builder.openArray();
     for (size_t i = 1; i <= nDocs; i++) {
+      builder.openArray();
+      builder.add(VPackSlice::nullSlice());
```
Why do we need the `nullSlice`?

markuspf commented on arangod/Aql/ModificationExecutorHelpers.h:
```diff
@@ -0,0 +1,99 @@
+// SmartGraphs in the enterprise edition. Refer to skipForAqlWrite in
+// Enterprise Edition
+bool writeRequired(ModificationExecutorInfos& infos, VPackSlice const& doc,
+                   std::string const& key);
```
Yes, if one also corrects the `BoolWrapper` in `ModificationExecutorInfos`. This is probably a site for further later beautification

markuspf commented on arangod/Aql/ModificationExecutor.cpp:
```diff
@@ -17,118 +17,202 @@
+  TRI_ASSERT(trx != nullptr);
+  bool const isDBServer = trx->state()->isDBServer();
+  _infos._producesResults = ProducesResults(
+      _infos._producesResults || (isDBServer && _infos._ignoreDocumentNotFound));
```
Thanks for pointing this out. This also removes the `TODO` above ;)

(I still put an explanation to where the code lives now...)

markuspf commented:
Thanks to @mchacki @goedderz and @ObiWahn for their reviews, I think I addressed all of the comments now  and started a jenkins PR test  and a performance test. 

https://jenkins01.arangodb.biz/view/PR/job/arangodb-matrix-pr/7050/

markuspf commented:
Tests were blue. Now of course a conflict again...

markuspf commented:
https://jenkins01.arangodb.biz/view/PR/job/arangodb-matrix-pr/7050/

markuspf commented:
https://jenkins01.arangodb.biz/view/PR/job/arangodb-matrix-pr/7063/
